### PR TITLE
PFORM-36: NUnit3 task supports NUnit2 for migration bridge

### PR DESCRIPTION
**Pull request by pnstickne**

### CODE REVIEW

File: Test/NUnit3.Tests.ps1 [hunk 1]
```diff
@@ -391,29 +448,37 @@ function ThenRanOnlySpecificTest
     }
 }
 
+function GetGeneratedNUnitReport {
```
Comment (splatteredbits) [inline Test/NUnit3.Tests.ps1]: Move this to the top with the other functions, ordered alphabetically. Rename to `Get-GeneratedNUnitReport1`.

File: Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1 [hunk 2]
```diff
@@ -29,8 +41,15 @@ function Invoke-WhiskeyNUnit3Task
         }
     }
 
-    $nunitReport = Join-Path -Path $TaskContext.OutputDirectory -ChildPath ('nunit3+{0}.xml' -f [IO.Path]::GetRandomFileName())
-    $nunitReportParam = '--result={0}' -f $nunitReport
+    $reportFormat = 'nunit3';
+    if ($TaskParameter['NUnitResultFormat'])
```
Comment (splatteredbits) [inline Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1]: Rename `NUnitResultFormat` to `ResultFormat`.

File: Test/NUnit3.Tests.ps1 [hunk 3]
```diff
@@ -44,22 +46,28 @@ function Init
...
+    $script:supportNUnit2 = $false
 
     $script:buildRoot = $TestDrive.FullName
 
     $script:outputDirectory = Join-Path -Path $buildRoot -ChildPath '.output'
 
-    Copy-Item -Path (Join-Path -Path $PSScriptRoot -ChildPath 'Assemblies\NUnit3*\bin\*\*') -Destination $buildRoot
+    # Test assemblies in separate folders to avoid cross-reference of NUnit Framework assembly versions
+    @(3, 2) | ForEach-Object  {
+        New-Item (Join-Path $buildRoot "NUnit$($_)Tests") -Type Directory
```
Comment (splatteredbits) [inline Test/NUnit3.Tests.ps1]: Do we need to add these to `.gitignore`?

File: Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1 [hunk 4]
```diff
@@ -16,7 +16,19 @@ function Invoke-WhiskeyNUnit3Task
     Set-StrictMode -Version 'Latest'
     Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
 
-    $nunitPackage = 'NUnit.ConsoleRunner'
+    $nunitPackages = @('NUnit.ConsoleRunner')
+    if ($TaskParameter['SupportNUnit2'])
+    {
+        # NUnit.Console includes NUnit.ConsoleRunner and the NUnit2 Compatibility Extension(s)
```
Comment (splatteredbits) [inline Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1]: Can you re-write this so it's a little clearer? I'm having a hard time following along.

File: Whiskey/en-US/about_Whiskey_NUnit3_Task.help.txt [hunk 5]
```diff
@@ -29,6 +29,8 @@ PROPERTIES
     * `OpenCoverArgument`: a list of additional arguments to pass to the OpenCover console.
     * `ReportGeneratorVersion`: the version of ReportGenerator to use. Defaults to the latest version available.
     * `ReportGeneratorArgument`: a list of additional arguments to pass to the ReportGenerator console.
+    * `SupportNUnit2`: boolean indicating if NUnit2 test support should be enabled. Defaults to false and is only intended for old non-migrated NUnit2 test projects.
+    * `NUnitResultFormat`: the result format of NUnit itself. See NUnit for supported formats. Defaults to 'nunit3'.
```
Comment (splatteredbits) [inline Whiskey/en-US/about_Whiskey_NUnit3_Task.help.txt]: Rename here, too.

File: Whiskey/en-US/about_Whiskey_NUnit3_Task.help.txt [hunk 6]
```diff
@@ -29,6 +29,8 @@ PROPERTIES
     * `OpenCoverArgument`: a list of additional arguments to pass to the OpenCover console.
     * `ReportGeneratorVersion`: the version of ReportGenerator to use. Defaults to the latest version available.
     * `ReportGeneratorArgument`: a list of additional arguments to pass to the ReportGenerator console.
+    * `SupportNUnit2`: boolean indicating if NUnit2 test support should be enabled. Defaults to false and is only intended for old non-migrated NUnit2 test projects.
```
Comment (splatteredbits) [inline Whiskey/en-US/about_Whiskey_NUnit3_Task.help.txt]: Add a note that using this option will download additional packages that include NUnit2 support.

File: Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1 [hunk 7]
```diff
@@ -16,7 +16,19 @@ function Invoke-WhiskeyNUnit3Task
     Set-StrictMode -Version 'Latest'
     Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
 
-    $nunitPackage = 'NUnit.ConsoleRunner'
+    $nunitPackages = @('NUnit.ConsoleRunner')
```
Comment (splatteredbits) [inline Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1]: Should we just *always* support NUnit2? Or always download the packages but select the runner differently based on the `SupportNUnit2` flag? How big are these packages?

File: Test/NUnit3.Tests.ps1 (hunk 3 above)
Comment (pnstickne) [reply Test/NUnit3.Tests.ps1]: Still copied to TestDrive - should not end up in repo path anywhere.

File: Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1 (hunk 7 above)
Comment (pnstickne) [reply Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1]: I'd be fine with that:

NUnit3 + 1 package, 1.2MB
1.2M    NUnit.ConsoleRunner.3.7.0

NUnit2 + 6 packages, ~500K.
24K     NUnit.Console.3.7.0
48K     NUnit.Extension.NUnitProjectLoader.3.5.0
333K    NUnit.Extension.NUnitV2Driver.3.6.0
52K     NUnit.Extension.NUnitV2ResultWriter.3.5.0
56K     NUnit.Extension.TeamCityEventListener.1.0.2
52K     NUnit.Extension.VSProjectLoader.3.5.0

Comment (splatteredbits) [reply Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1]: Let's do that then. Always download all the packages but select the "right" one when we need to support NUnit2.

File: Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1 [hunk 8]
```diff
@@ -16,18 +16,10 @@ function Invoke-WhiskeyNUnit3Task
...
-        # NUnit3 only
-        $nunitPackages = @('NUnit.Console', 'NUnit.ConsoleRunner')
-    }
+    # NUnit.Console pulls in ConsoleRunner (of the same version) as a dependency and several NUnit2 compatibility/extension packages.
+    # The ConsoleRunner packages is installed explicitly to resolve the tool/bin path from installed package location.
+    $nunitSupportPackage = 'NUnit.Console'
+    $nunitPackage = 'NUnit.ConsoleRunner'
 
     # Due to a bug in NuGet we can't search for and install packages with wildcards (e.g. 3.*), so we're hardcoding a version for now. See Resolve-WhiskeyNuGetPackageVersion for more details.
     $nunitVersion = '3.7.0'
```
Comment (splatteredbits) [inline Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1]: Maybe we should upgrade to latest NUnit 3 while we're at it? 3.12. I didn't realize in my earlier review we were pinned to a specific version.

Comment (pnstickne) [reply Whiskey/Tasks/Invoke-WhiskeyNUnit3Task.ps1]: Updated to 3.10 (latest version of https://www.nuget.org/packages/NUnit.ConsoleRunner/ ; core library versioning is close-yet-independent.)

**Pull request merged**